File: frontend/src/api/tester-agent.ts
/**
 * Tester Agent API - PRD P4-002
 *
 * The Tester Agent takes a requirement (with optional acceptance criteria),
 * generates test cases, executes them, and reports any bugs found as work
 * items. Each run is tracked as a TesterJob with a full lifecycle:
 * pending → generating_cases → executing → reporting → completed.
 */
import api from './index'

export type TesterJobStatus =
  | 'pending'
  | 'generating_cases'
  | 'executing'
  | 'reporting'
  | 'completed'
  | 'failed'
  | 'cancelled'

export interface TestCase {
  id: string
  name: string
  description: string
  steps: string[]
  expected: string
}

export interface TestResult {
  case_id: string
  name: string
  status: 'passed' | 'failed' | 'skipped'
  duration_ms: number
  error?: string
}

export interface TesterJob {
  id: number
  workspace_id: number
  project_id?: number
  issue_id?: number
  agent_task_id?: number
  title: string
  requirement_text: string
  acceptance_criteria: string
  test_scope: string // unit | integration | e2e
  input_context: Record<string, unknown>
  generated_cases: TestCase[]
  test_results: TestResult[]
  total_cases: number
  pass_count: number
  fail_count: number
  skip_count: number
  bug_issue_ids: number[]
  status: TesterJobStatus
  progress: number
  current_step?: string
  error_message?: string
  started_at?: string
  completed_at?: string
  cancelled_at?: string
  created_at: string
  updated_at: string
}

export interface TesterJobCreate {
  title: string
  requirement_text?: string
  acceptance_criteria?: string
  test_scope?: string
  project_id?: number
  issue_id?: number
  cases?: TestCase[]
  input_context?: Record<string, unknown>
}

export interface TesterJobListFilter {
  status?: TesterJobStatus
  limit?: number
}

export const testerAgentApi = {
  list(workspaceId: number, filter: TesterJobListFilter = {}): Promise<TesterJob[]> {
    const params: Record<string, string | number> = {}
    if (filter.status) params.status = filter.status
    if (filter.limit) params.limit = filter.limit
    return api
      .get(`/workspaces/${workspaceId}/tester-agent/jobs`, { params })
      .then(res => res.data)
  },

  get(workspaceId: number, jobId: number): Promise<TesterJob> {
    return api
      .get(`/workspaces/${workspaceId}/tester-agent/jobs/${jobId}`)
      .then(res => res.data)
  },

  create(workspaceId: number, payload: TesterJobCreate): Promise<TesterJob> {
    return api
      .post(`/workspaces/${workspaceId}/tester-agent/jobs`, payload)
      .then(res => res.data)
  },

  cancel(workspaceId: number, jobId: number): Promise<TesterJob> {
    return api
      .post(`/workspaces/${workspaceId}/tester-agent/jobs/${jobId}/cancel`)
      .then(res => res.data)
  },

  delete(workspaceId: number, jobId: number): Promise<void> {
    return api
      .delete(`/workspaces/${workspaceId}/tester-agent/jobs/${jobId}`)
      .then(() => undefined)
  }
}

File: backend/internal/model/tester_agent.go
package model

import (
	"encoding/json"
	"time"
)

// TesterJobStatus defines the lifecycle state of a Tester Agent job.
//
// Flow: pending → generating_cases → executing → reporting → completed
// Any step may transition to "failed".
type TesterJobStatus string

const (
	TesterJobPending         TesterJobStatus = "pending"
	TesterJobGeneratingCases TesterJobStatus = "generating_cases"
	TesterJobExecuting       TesterJobStatus = "executing"
	TesterJobReporting       TesterJobStatus = "reporting"
	TesterJobCompleted       TesterJobStatus = "completed"
	TesterJobFailed          TesterJobStatus = "failed"
	TesterJobCancelled       TesterJobStatus = "cancelled"
)

// TesterJob tracks a single Tester Agent execution (PRD P4-002).
//
// The Tester Agent generates test cases from a requirement + acceptance
// criteria, executes them, and reports failures as Bug work items. Each job
// records the generated cases, execution results, and the IDs of any Bug
// issues created so the lifecycle can be observed and audited from the UI.
type TesterJob struct {
	BaseModel

	WorkspaceID uint64  `gorm:"not null;index" json:"workspace_id"`
	ProjectID   *uint64 `gorm:"index" json:"project_id,omitempty"`
	IssueID     *uint64 `gorm:"index" json:"issue_id,omitempty"`
	AgentTaskID *uint64 `gorm:"index" json:"agent_task_id,omitempty"`

	// Inputs
	Title              string          `gorm:"size:255;not null" json:"title"`
	RequirementText    string          `gorm:"type:text" json:"requirement_text"`
	AcceptanceCriteria string          `gorm:"type:text" json:"acceptance_criteria"`
	TestScope          string          `gorm:"size:20;default:unit" json:"test_scope"` // unit | integration | e2e
	InputContext       json.RawMessage `gorm:"type:jsonb;default:'{}'" json:"input_context"`

	// Generated test cases: array of {id, name, description, steps, expected}
	GeneratedCases json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"generated_cases"`

	// Execution results: array of {case_id, name, status, duration_ms, error}
	TestResults json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"test_results"`

	// Roll-up summary
	TotalCases int `gorm:"default:0" json:"total_cases"`
	PassCount  int `gorm:"default:0" json:"pass_count"`
	FailCount  int `gorm:"default:0" json:"fail_count"`
	SkipCount  int `gorm:"default:0" json:"skip_count"`

	// Bug issue IDs created from failed cases (linked as sub-issues of IssueID)
	BugIssueIDs json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"bug_issue_ids"`

	// Lifecycle
	Status       TesterJobStatus `gorm:"size:20;default:pending;index" json:"status"`
	Progress     int             `gorm:"default:0" json:"progress"`
	CurrentStep  *string         `gorm:"size:100" json:"current_step,omitempty"`
	ErrorMessage *string         `gorm:"type:text" json:"error_message,omitempty"`

	StartedAt   *time.Time `json:"started_at,omitempty"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
	CancelledAt *time.Time `json:"cancelled_at,omitempty"`

	// Relationships
	Workspace Workspace `gorm:"foreignKey:WorkspaceID" json:"-"`
	Project   *Project  `gorm:"foreignKey:ProjectID" json:"-"`
	Issue     *Issue    `gorm:"foreignKey:IssueID" json:"-"`
}

func (TesterJob) TableName() string { return "tester_jobs" }
